direct/ingest-server/src/main/java/cz/o2/proxima/server/ReplicationController.java

```diff
@@ -0,0 +1,409 @@
+        log.debug(
+            "Consumer {}: writing element {} into {}",
+            consumerName, ingest, writer);
+        writer.write(ingest, (success, exc) -> confirmWrite(
+            consumerName, ingest, writer, success, exc, context::confirm, context::fail));
+      }
+    };
+    return RetryableLogObserver.online(3, consumerName, commitLog, logObserver);
+  }
+
+  private void confirmWrite(
+      String consumerName,
+      StreamElement ingest,
+      AttributeWriterBase writer,
+      boolean success, Throwable exc,
+      Runnable onSuccess,
+      Consumer<Throwable> onError) {
+
+    if (!success) {
+      log.error(
+          "Consumer {}: failed to write ingest {} to {}",
+          consumerName, ingest, writer.getUri(), exc);
+      Metrics.NON_COMMIT_WRITES_RETRIES.increment();
+      if (ignoreErrors) {
+        log.error(
+            "Consumer {}: retries exhausted trying to ingest {} to {}. "
+                + "Configured to ignore. Skipping.",
+            consumerName, ingest, writer.getUri());
+        onSuccess.run();
+      } else {
+        onError.accept(exc);
+      }
+    } else {
+      if (ingest.isDelete()) {
+        Metrics.NON_COMMIT_LOG_DELETES.increment();
+      } else {
+        Metrics.NON_COMMIT_LOG_UPDATES.increment();
+      }
+      onSuccess.run();
+    }
+  }
+
+  private void onReplicationError(Throwable t) {
+
+  }
+
+  @AllArgsConstructor
+  private abstract class ReplicationLogObserver implements LogObserver {
+
+    private final String consumerName;
+    private final CommitLogReader commitLog;
+    private final Set<AttributeDescriptor<?>> allowedAttributes;
+    private final StorageFilter filter;
+    private final AttributeWriterBase writer;
+
+    @Override
+    public boolean onNext(
+        StreamElement ingest,
+        OnNextContext context) {
+      final boolean allowed = allowedAttributes.contains(
+          ingest.getAttributeDescriptor());
+      log.debug(
+          "Consumer {}: received new ingest element {}", consumerName, ingest);
+      if (allowed && filter.apply(ingest)) {
+        Failsafe.with(retryPolicy).run(() -> ingestElement(ingest, context));
+      } else {
+        Metrics.COMMIT_UPDATE_DISCARDED.increment();
+        log.debug(
+            "Consumer {}: discarding write of {} to {} because of {}, "
+                + "with allowedAttributes {} and filter class {}",
+            consumerName, ingest, writer.getUri(),
+            allowed ? "applied filter" : "invalid attribute",
+            allowedAttributes, filter.getClass());
+      }
+      return true;
+    }
+
+    @Override
+    public boolean onError(Throwable error) {
+      onReplicationError(
+          new IllegalStateException(
+              String.format(
+                  "Consumer %s: too many errors retrying the consumption of commit log %s.",
+                  consumerName, commitLog.getUri())));
+      return false;
```

Comment:
This is definitely wrong. We cannot do nothing (onReplicationError) and return `false`, because that would mean to stop consumption. That way - every error during consumption of the commit log would lead to replication being stopped.